### PR TITLE
177 RST stubs for ReadTheDocs

### PR DESCRIPTION
**Pull request by bbean23**
## Purpose

_Please briefly describe the purpose of this pull request. If this fixes a github issue please include `e.g. fixes #22` so that the issue gets closed automatically when this is merged._

## Summary of changes

- provided RST stubs for SpotAnalysis classes
- updated docstrings
- added classes to test_DocStringsExist.py

## Implementation notes

Added AutoSummary to Sphinx for generating class name lists at the top of the documentation files.

## Submission checklist
- [x] Target branch is `develop`, not `main`
- [x] Existing tests are updated or new tests were added
- [x] `opencsp/test/test_DocStringsExist.py` are verified to include this change or have been updated accordingly
- [x] .rst file(s) under `doc/` are verified to include this change or have been updated accordingly

## Additional information

none

### CODE REVIEW

File: opencsp/common/lib/cv/spot_analysis/SpotAnalysisOperablesStream.py [hunk 1]
```diff
@@ -4,20 +4,43 @@
...
+            operables. This stream will be restartable as long as the given
+            'images' stream is restartable.
+        """
         self.images = images
         self.images_iter = None
         self.default_support_images: dict[ImageType, CacheableImage] = None
         self.default_data: SpotAnalysisOperable = None
 
     def set_defaults(self, default_support_images: dict[ImageType, CacheableImage], default_data: SpotAnalysisOperable):
+        """
```
Comment (e10harvey) [inline opencsp/common/lib/cv/spot_analysis/SpotAnalysisOperablesStream.py:37]: Missing parameter documentation in doc string

File: doc/source/library_reference/library/SpotAnalysis/base_classes.rst [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+These are the book keeping classes that enable the spot analysis pipeline.
+
+.. currentmodule:: opencsp.common.lib.cv.spot_analysis
+
+.. automodule:: opencsp.common.lib.cv.spot_analysis.SpotAnalysisOperable
+   :members:
+   :undoc-members:
+   :show-inheritance:
+   :member-order: groupwise
+
```
Comment (e10harvey) [inline doc/source/library_reference/library/SpotAnalysis/base_classes.rst:25]:
```suggestion

.. automodule:: opencsp.common.lib.cv.spot_analysis.SpotAnalysisOperablesStream
   :members:
   :undoc-members:
   :show-inheritance:
   :member-order: groupwise

```

File: opencsp/common/lib/cv/spot_analysis/image_processor/FalseColorImageProcessor.py [hunk 3]
```diff
@@ -36,6 +36,12 @@ def __init__(self, map_type='human', opencv_map=cv2.COLORMAP_JET):
         self.opencv_map = opencv_map
 
     def apply_mapping_jet_custom(self, operable: SpotAnalysisOperable) -> SpotAnalysisOperable:
+        """Updates the primary image with a false color map ('human' or
```
Comment (e10harvey) [inline opencsp/common/lib/cv/spot_analysis/image_processor/FalseColorImageProcessor.py:39]: Missing parameter docs for operable

File: doc/source/library_reference/library/SpotAnalysis/image_processor.rst [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+processing/machine vision situation. 
+
+.. currentmodule:: opencsp.common.lib.cv.spot_analysis.image_processor
+
+.. autosummary::
+   :nosignatures:
+
+   AbstractAggregateImageProcessor
+   AbstractSpotAnalysisImageProcessor
+   AbstractVisualizationImageProcessor
```
Comment (e10harvey) [inline doc/source/library_reference/library/SpotAnalysis/image_processor.rst:16]:
```suggestion
   AbstractSpotAnalysisImageProcessorLeger
   AbstractVisualizationImageProcessor
```

File: opencsp/test/test_DocStringsExist.py [hunk 5]
```diff
@@ -1,28 +1,70 @@
...
+                if inspect.getdoc(getattr(class_module, method)) is None:
+                    doc_exists = False
+
+                method_name = f"{class_module.__name__}.{method}"
+                print(f"doc_exists({method_name}): " f"{doc_exists}")
+                if not doc_exists:
+                    undocumented_methods.append(method)
+
+            self.assertEqual(
+                len(undocumented_methods), 0, "Found undocumented methods:\n\t" + "\n\t".join(undocumented_methods)
```
Comment (e10harvey) [inline opencsp/test/test_DocStringsExist.py:65]: thank you

**Review by e10harvey — CHANGES_REQUESTED** [review]
Thank you for these changes @bbean23. I left some minor comments below. Please reorganize these rst files to match the directory structore of the code base, like this:
`doc/source/library_reference/library/SpotAnalysis -> doc/source/library_reference/common/lib/cv/spot_analysis`. I'd be happy to open a PR into your fork with these changes.

**Review by bbean23 — COMMENTED** [review]
I believe I've addressed all your comments. Please note that by "addressed" I sometimes mean "argued against", so please take a look at my comments to see if you agree or not.

File: doc/source/library_reference/library/SpotAnalysis/base_classes.rst [hunk 6]
```diff
@@ -0,0 +1,52 @@
+SpotAnalysis
+============
```
[the remark this answers is not in the document]
Comment (e10harvey) [reply doc/source/library_reference/library/SpotAnalysis/base_classes.rst]: I don't have a strong preference but the current documentation uses the second style

File: doc/source/library_reference/library/SpotAnalysis/index.rst [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+Spot Analysis is the generalized analysis of a beam of light on an optical
+target. The code for this is set up as a pipelined image processing system where
+images are passed through a set of image processors that reduce/infer
+information from those images.
+
+.. toctree::
+   :maxdepth: 1
+
+   base_classes.rst
+   image_processor.rst
```
[the remark this answers is not in the document]
Comment (e10harvey) [reply doc/source/library_reference/library/SpotAnalysis/index.rst]: Either way is fine with me but I want to make sure both the code and rst is maintainable and organized. Maybe the compromise is appending `config_` to base_classes.rst and image_processor.rst.

File: doc/source/library_reference/library/SpotAnalysis/base_classes.rst (hunk 2 above)
Comment (e10harvey) [reply doc/source/library_reference/library/SpotAnalysis/base_classes.rst]: I would be inclined to include all non-protected and non-private entities in the documentation

**Review by e10harvey — COMMENTED** [review]
Thank you @bbean23! I think this can be merged after the headings are made consistent with the rest of the library reference.

File: doc/source/library_reference/library/SpotAnalysis/index.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (bbean23) [reply doc/source/library_reference/library/SpotAnalysis/index.rst]: I'm not sure what you mean by this. You mean changing the file name to `base_classes_config.rst` and `image_processor_config.rst`?

File: opencsp/common/lib/cv/SpotAnalysis.py [hunk 8]
```diff
@@ -34,85 +35,117 @@ class SpotAnalysis(Iterator[tuple[SpotAnalysisOperable]]):
...
+    A list of possible use cases for this class include:
+        a. BCS
+        b. BCS-based heliostat calibration
+        c. Vertical laser characterization
+        d. Laser beam spread characterization
+        e. 2f returned spot analysis
+        f. Motion characterization video analysis
+        g. Wind BCS video trajectory analysis
+        h. Laser-based cross-check
+        i. Earth mover's metric for comparisons (scipy has a method, also maaaaaybe look at https://www.cs.cmu.edu/~efros/courses/LBMV07/Papers/rubner-jcviu-00.pdf)
```
Comment (e10harvey) [inline opencsp/common/lib/cv/SpotAnalysis.py:47]:
```suggestion
        i. Earth mover's metric for comparisons (scipy has a method, also look at https://www.cs.cmu.edu/~efros/courses/LBMV07/Papers/rubner-jcviu-00.pdf)
```

**Review by e10harvey — APPROVED** [review]

File: doc/source/library_reference/library/SpotAnalysis/index.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (bbean23) [reply doc/source/library_reference/library/SpotAnalysis/index.rst]: To be addressed in follow-on issue

File: doc/source/library_reference/library/SpotAnalysis/base_classes.rst (hunk 2 above)
Comment (bbean23) [reply doc/source/library_reference/library/SpotAnalysis/base_classes.rst]: Todo: indicate that this class is not meant for use by users, and most of the time not for use by developers. It is necessary for how the SpotAnalysis class works but only users who really want a deep dive should take a look. Maybe _SpotAnalysisOperablesStream or __SpotAnalysisOperablesStream?

Comment (bbean23) [reply doc/source/library_reference/library/SpotAnalysis/base_classes.rst]: Done!

File: opencsp/common/lib/cv/SpotAnalysis.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (bbean23) [reply opencsp/common/lib/cv/SpotAnalysis.py]: Thanks for your fix on this @e10harvey !

**Comment by bbean23** [comment]
@e10harvey Please review again. I made some small changes in order to get the build to pass and documentation to render correctly:

* fixed docstrings in CacheableImage to be compatible with ReStructuredText
* changed "AbstractSpotAnalysisImagesProcessor" to "AbstractSpotAnalysisImageProcessor"

**Pull request merged**